compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3850,7 +3850,56 @@ TR::Register *OMR::Power::TreeEvaluator::vdsqrtEvaluator(TR::Node *node, TR::Cod
 
 TR::Register* OMR::Power::TreeEvaluator::vfmaEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    {
-   return TR::TreeEvaluator::unImpOpEvaluator(node, cg);
+   TR::DataType type = node->getDataType().getVectorElementType();
+   if (type != TR::Float && type != TR::Double)
+   {
+      TR_ASSERT(false, "unsupported vector type %s\n", type.toString());
+      return NULL;
+   }
+
+   TR::InstOpCode::Mnemonic op;
+
+   TR::Node *firstChild  = node->getFirstChild();
+   TR::Node *secondChild = node->getSecondChild();
+   TR::Node *thirdChild  = node->getThirdChild();
+
+   TR::Register *thirdReg  = cg->evaluate(thirdChild);
+   TR::Register *firstReg  = cg->evaluate(firstChild);
+   TR::Register *secondReg = cg->evaluate(secondChild);
+
+   TR::Register *resReg;
+
+   if (cg->canClobberNodesRegister(thirdChild))
+      {
+      resReg = thirdReg;
+      op = (type == TR::Float) ? TR::InstOpCode::xvmaddasp : TR::InstOpCode::xvmaddadp;
+      generateTrg1Src2Instruction(cg, op, node, thirdReg, firstReg, secondReg);
+      }
+   else if (cg->canClobberNodesRegister(firstChild))
+      {
+      resReg = firstReg;
+      op = (type == TR::Float) ? TR::InstOpCode::xvmaddmsp : TR::InstOpCode::xvmaddmdp;
+      generateTrg1Src2Instruction(cg, op, node, firstReg, secondReg, thirdReg);
+      }
+   else if (cg->canClobberNodesRegister(secondChild))
+      {
+      resReg = secondReg;
+      op = (type == TR::Float) ? TR::InstOpCode::xvmaddmsp : TR::InstOpCode::xvmaddmdp;
+      generateTrg1Src2Instruction(cg, op, node, secondReg, firstReg, thirdReg);
+      }
+   else
+      {
+      resReg = cg->allocateRegister(TR_VSX_VECTOR);
```

Comment:
A comment here saying that all of the operand registers are still needed later so the code needs to copy one of the operands for this to work.